include/clap/preset-discovery.h

```diff
@@ -0,0 +1,244 @@
+   const char        *id;
+   const char        *name;
+   const char        *vendor;
+   const char *const *plugin_ids; // null terminated array of plugin ids
+} clap_preset_provider_descriptor_t;
+
+// This interface isn't thread-safe.
+typedef struct clap_preset_provider {
+   const clap_preset_provider_descriptor_t *desc;
+
+   // Destroys the preset provider
+   void(CLAP_ABI *destroy)(const struct clap_preset_provider *provider);
+
+   // Retrives the path to a watch file.
+   // Whenever the given file is "touched", then the indexer shall invalidate all the data.
+   bool(CLAP_ABI *invalidation_watch_file)(const struct clap_preset_provider *provider,
+                                           char                              *watch_file_path,
+                                           uint32_t watch_file_path_capacity);
+
+   // returns the number of locations
+   uint32_t(CLAP_ABI *locations_count)(const struct clap_preset_provider *provider);
+
+   // gets the location info at index
+   bool(CLAP_ABI *get_location)(const struct clap_preset_provider *provider,
+                                uint32_t                           index,
+                                clap_preset_location_t            *out_info);
+
+   // returns the number of different supported file extensions
+   uint32_t(CLAP_ABI *file_extensions_count)(const struct clap_preset_provider *provider);
+
+   // stores the file extensions at index into out_extension.
+   // the `.' isn't included in the string.
+   // extension_capacity indicates the number of bytes available to store the extension
+   bool(CLAP_ABI *get_file_extension)(const struct clap_preset_provider *provider,
+                                      uint32_t                           index,
+                                      char                              *out_extension,
+                                      uint32_t                           extension_capacity);
+
+   // gets information about a given collection.
+   bool(CLAP_ABI *get_collection_info)(const struct clap_preset_provider *provider,
+                                       const char                        *collection_id,
+                                       clap_preset_collection_info_t     *out_info);
+
+   // reads metadata from the given file and passes them to the metadata receiver
+   bool(CLAP_ABI *read_file_metadata)(const struct clap_preset_provider     *provider,
+                                      const char                            *file_path,
+                                      const clap_preset_metadata_receiver_t *metadata_receiver);
+} clap_preset_provider_t;
+
+typedef struct clap_preset_indexer {
+   clap_version_t clap_version; // initialized to CLAP_VERSION
+   const char    *name;
+   const char    *vendor;
+   const char    *version;
+} clap_preset_indexer_t;
+
+static const CLAP_CONSTEXPR char CLAP_PRESET_DISCOVERY_FACTORY_ID[] =
+   "clap.preset-discovery-factory";
```

Comment:
This definition should probably be moved to the top of the file for consistency and to make it easier to find. Same for the othe factories where this is not yet the case.

Reply:
On the other hand, the factory id is next to the struct, so they're kind "connected" that way.

Reply:
They're kind of difficult to find now. I've had to point multiple people to them because they never saw them looking at the header and in turn assumed the factory ID string was just some placeholder they could ignore.